### PR TITLE
Add missing constructors for `Bidiagonal`, `Symmetric`, and `Hermitian`

### PR DESCRIPTION
**Pull request by mcognetta**
`Bidiagonal`, `Symmetric`, and `Hermitian` are currently all missing the `X(A::Matrix, uplo::Char)` constructor but have the  `X(A::Matrix, uplo::Symbol)` constructor. This PR adds the `Char` constructor.

Two additional methods `Symmetric(A::AbstractArray) = Symmetric(A, :U)` and `Hermitian(A::AbstractMatrix) = Hermitian(A, :U)` were also added to avoid the method overwriting warning (due to the previous implementation having default arguments) and to preserve the historical default argument.

**NOTE:** This PR should be submitted after #42466, which adds better `uplo` checks to the constructors.

-------------------------------------------------------------------------------

HEAD:

```julia
julia> Bidiagonal(rand(3,3), :U)
3×3 Bidiagonal{Float64,Array{Float64,1}}:
 0.971135  0.423251   ⋅ 
  ⋅        0.528393  0.696667
  ⋅         ⋅        0.343728

julia> Bidiagonal(rand(3,3), 'U')
ERROR: MethodError: no method matching Bidiagonal(::Array{Float64,2}, ::Char)
Closest candidates are:
  Bidiagonal(::AbstractArray{T,2} where T, ::Symbol) at /home/mc/github/julia/usr/share/julia/stdlib/v1.5/LinearAlgebra/src/bidiag.jl:101
Stacktrace:
 [1] top-level scope at REPL[3]:1
```

This PR:

```julia
julia> Bidiagonal(rand(3,3), :U)
3×3 Bidiagonal{Float64, Vector{Float64}}:
 0.87165  0.860882   ⋅ 
  ⋅       0.539619  0.278167
  ⋅        ⋅        0.4687

julia> Bidiagonal(rand(3,3), 'U')
3×3 Bidiagonal{Float64, Vector{Float64}}:
 0.499547  0.862715   ⋅ 
  ⋅        0.556194  0.751918
  ⋅         ⋅        0.706695
```

and likewise for `Symmetric`, and `Hermitian`.

### CODE REVIEW

**Comment by mcognetta** [comment]
Failures look unrelated.

**Comment by N5N3** [comment]
1. If you want to introduce a `Char` constructor to `Hermitian`, you'd better to add it to `hermitian` to keep consistency.
2. `Symbol` constructor checks `:L`, `:U` via `char_uplo`, so I think we should do similar check in `Char` constructor.
3. I also wonder why you say it's missing. Can you give an example that `Char` constructor is better than `Symbol` constructor.

File: stdlib/LinearAlgebra/test/bidiag.jl [hunk 1]
```diff
@@ -48,6 +48,10 @@ Random.seed!(1)
             # from matrix
             @test Bidiagonal(ubd, :U) == Bidiagonal(Matrix(ubd), :U) == ubd
             @test Bidiagonal(lbd, :L) == Bidiagonal(Matrix(lbd), :L) == lbd
+
+            z = rand(elty, 3, 3)
+            @test Bidiagonal(z, 'U') === Bidiagonal(z, :U)
```
Comment (N5N3) [inline stdlib/LinearAlgebra/test/bidiag.jl:53]: This test wont pass as `Bidiagonal` allocates 2 Vector to store the diag and upper/lower diag. 
Replace `===` with `==`?

Comment (mcognetta) [reply stdlib/LinearAlgebra/test/bidiag.jl:53]: Done. Thanks for the catch.

**Comment by mcognetta** [comment]
> 1. If `Char` constructor is introduced to `Hermitian`, you'd better to add it to `hermitian` to keep consistency.

Done. Actually, this and `symmetric` aren't exported. Perhaps they should be.

> 2. `Symbol` constructor checks `:L`, `:U` via `char_uplo`, so I think we should do similar check in `Char` constructor.

This is done in #42466. I will add a note not to submit this one until that is merged.

> 3. As for `Symmetric(A::AbstractArray) = Symmetric(A, :U)`, since
> 
> ```julia
> julia> f(a,b::Int = 1) = 1
> f (generic function with 2 methods)
> julia> f(a,b::Float64) = 2
> f (generic function with 3 methods)
> julia> f(1,1), f(1), f(1,1.0)
> (1, 1, 2)
> ```
> I think you only need to add the new constructor?

Doing it this way causes a Method Overwriting warning during compilation (doesn't happen in the REPL I guess, but it happens when running `make` for example.). This discusses it (but, at the time, the warning would still appear in REPL. the same code does not produce the warning in REPL at HEAD):
https://discourse.julialang.org/t/method-definition-overwritten-warning-for-functions-with-default-arguments/1710

> 4. I also wonder why you say it's missing. Can you give us an example that `Char` constructor is better than `Symbol` constructor?

It's missing in the sense that some of the constructors allow `Symbol` and `Char` but some don't. I don't think `Char` is better, but adding those constructors makes the API more consistent. 

---------------------------------------------------------------------------------------

Thanks for the comments.

**Comment by mcognetta** [comment]
Note that this should be submitted after #42466, so that the correct `uplo` checks are included in the base constructors.

**Comment by dkarrasch** [comment]
You can rebase onto current master now.

File: stdlib/LinearAlgebra/src/symmetric.jl [hunk 2]
```diff
@@ -126,16 +137,25 @@ All non-real parts of the diagonal will be ignored.
...
 """
-function Hermitian(A::AbstractMatrix, uplo::Symbol=:U)
+function Hermitian(A::AbstractMatrix, uplo::Symbol)
     n = checksquare(A)
     return hermitian_type(typeof(A))(A, char_uplo(uplo))
 end
 
+# suppress method overwritten warning while preserving
+# the historical default argument `uplo=:U`
+Hermitian(A::AbstractMatrix) = Hermitian(A, :U)
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/symmetric.jl:147]: Is this a new issue or has it always been there but we ignored the warning?

**Comment by dkarrasch** [comment]
I'm not sure, but I think not exporting `symmetric` and `hermitian` was done on purpose. Unless we have a good reason, I wouldn't export them, because you can do what you want by calling the wrapper constructors with capitalized initial letters.

**Comment by andreasnoack** [comment]
It's on purpose that we don't have `Char` constructors. The idea was that `Char` was just the internal representation whereas `Symbol` was the public API.

**Comment by mcognetta** [comment]
> It's on purpose that we don't have `Char` constructors. The idea was that `Char` was just the internal representation whereas `Symbol` was the public API.

Oh, I see. In that case, should I close this PR?

**Pull request closed without merging**